compiler/rustc_mir_dataflow/src/framework/direction.rs

```diff
@@ -286,6 +306,33 @@ impl Direction for Backward {
     }
 }
 
+struct BackwardSwitchIntEdgeEffectApplier<'a, D, F> {
+    pred: BasicBlock,
+    exit_state: &'a mut D,
+    targets: &'a SwitchTargets,
+    bb: BasicBlock,
+    propagate: &'a mut F,
+
+    effects_applied: bool,
+}
+
+impl<D, F> super::SwitchIntEdgeEffect<D> for BackwardSwitchIntEdgeEffectApplier<'_, D, F>
+where
+    D: Clone,
+    F: FnMut(BasicBlock, &D),
+{
+    fn apply(&mut self, mut apply_edge_effect: impl FnMut(&mut D, SwitchIntTarget)) {
+        assert!(!self.effects_applied);
+
+        let value =
+            self.targets.iter().find_map(|(value, target)| (target == self.bb).then_some(value));
```

Comment:
I don't like signing off on quadratic algorithms when alternatives exist. Adding a lazily computed data structure to `Body` (a lá [`predecessors`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/mir/predecessors.rs)) seems like the easiest way. Ideally this would be part of a generic graph adapter for the MIR CFG that supports efficient lookup of edge data, but that's outside the scope of this PR.

Reply:
I see now there's a bug here: this code incorrectly assumes that only one value transfers control from `pred` to `target`. Please expect to see this fixed in a subsequent push.